taxcalc/parameters.py

```diff
@@ -1,170 +1,496 @@
-                           " values for the years between {0} and {1}.")
-                    raise ValueError(msg.format(self.current_year, year))
-
-            # Set up the '_X = [a, b,...]' variables as 'X = a'
-            self.set_year(self._current_year)
+        if self.current_year != self.start_year:
+            self.set_year(self.start_year)
+        if not reform:
+            return  # no reform to implement
+        reform_years = sorted(list(reform.keys()))
+        last_reform_year = max(reform_years)
+        if last_reform_year > self.end_year:
+            msg = 'reform provision in year={} > end_year={}'
+            ValueError(msg.format(last_reform_year, self.end_year))
+        for year in reform_years:
+            if year != self.start_year:
+                self.set_year(year)
+            self._update({year: reform[year]})
+        self.set_year(self.start_year)
 
     @property
     def current_year(self):
+        """
+        Current policy parameter year property.
+        """
         return self._current_year
 
     @property
     def start_year(self):
+        """
+        First policy parameter year property.
+        """
         return self._start_year
 
     @property
-    def budget_years(self):
-        return self._budget_years
+    def num_years(self):
+        """
+        Number of policy parameter years property.
+        """
+        return self._num_years
 
-    def increment_year(self):
-        self._current_year += 1
-        self.set_year(self._current_year)
+    @property
+    def end_year(self):
+        """
+        Last policy parameter year property.
+        """
+        return self._end_year
 
-    def set_year(self, yr):
-        for name, vals in self._vals.items():
+    def set_year(self, year):
+        """
+        Set policy parameters to values for specified calendar year.
+
+        Parameters
+        ----------
+        year: int
+            calendar year for which to current_year and parameter values
+
+        Raises
+        ------
+        ValueError:
+            if year is not in [start_year, end_year] range.
+
+        Returns
+        -------
+        nothing: void
+
+        Notes
+        -----
+        To increment the current year, use the following statement::
+
+            ppo.set_year(ppo.current_year + 1)
+
+        where ppo is a policy Parameters object.
+        """
+        if year < self.start_year or year > self.end_year:
+            msg = 'year passed to set_year() must be in [{},{}] range.'
+            raise ValueError(msg.format(self.start_year, self.end_year))
+        self._current_year = year
+        year_zero_indexed = year - self._start_year
+        for name in self._vals:
             arr = getattr(self, name)
-            setattr(self, name[1:], arr[yr - self._start_year])
+            setattr(self, name[1:], arr[year_zero_indexed])
+
+    @classmethod
+    def default_data(cls, metadata=False, start_year=None):
```

Comment:
Maybe we should schedule a code review to increase clarity. There are three repositories in play here. There is this one (Tax-Calculator), dropq, and webapp-public. Last week, I reported that the new class method `default_data` passes a test in `webapp-public`. `webapp-public` is a user of both the `taxcalc` package and the `dropq` package. `dropq` is a user of the `taxcalc` package. In order to fully run the web application (and not just the unit tests in `webapp-public`, I attempted to run the `dropq` package in a way similar to how it is done in `webapp-public`. I was unable to do this for the above stated reasons. I did not post code that I was unable to get work; I posted code that illustrated the problem with the code in this branch. The code in `dropq` that fails when run against this branch is best illustrated by looking at `test_make_Calculator_increment_years_first`. If this test were to pass as it exists in master, then the `dropq` package would work, because the test is designed to show how `dropq` uses the package. It seems more helpful to focus on a test or show a small snippet of code indicating the failure in assumptions than to say something like: 

Run this file:
https://github.com/OpenSourcePolicyCenter/dropQ/blob/master/example.py

And you'll get an assertion violation at this line in `calculate.py`:

```
 assert self._params.current_year == self._records.current_year
```

That would send someone down a path of digging in to a bunch of code attempting to discover why the assertion was violated. My goal was to save time and get to the heart of the issue, which for this method, is a mutation of `current_year` where there previously had not been one.